src/core/utils.js

```diff
@@ -105,6 +106,56 @@ export function loadAddonIfNeeded(
     .then(({ entities }) => dispatch(loadEntities(entities)));
 }
 
+export function fetchVersionDetails(
+  { store: { dispatch, getState }, params: { slug } }
+) {
+  const state = getState();
+  const addon = findAddon(state, slug);
+
+  if (!addon) {
+    return null;
+  }
+  dispatch(versionGet({ slug, versionID: addon.current_version.id }));
+
+  return versionDetail({
+    api: state.api,
+    slug: addon.slug,
+    versionID: addon.current_version.id,
+  })
+    .then((fetchedVersion) => dispatch(versionLoad(fetchedVersion)))
+    .catch((response) => dispatch(versionFail(response)));
+}
+
+export function loadAddonAndVersionDetail(
+  { store: { dispatch, getState }, params: { slug } }
+) {
+  const state = getState();
+  const addon = loadAddonIfNeeded({
+    store: { dispatch, getState },
+    params: { slug },
+  });
+
+  if (!addon.current_version) {
+    return addon
+      .then(() => fetchVersionDetails({
+        store: { dispatch, getState },
+        params: { slug },
+      }));
+  }
+
+  // Check for this add-on's current version details in the existing state.
+  if (state.version && state.version[addon.current_version.id]) {
+    dispatch(versionGet({ slug, versionID: addon.current_version.id }));
+    log.info(`Found version ${addon.current_version.id} for addon ${addon.id} in state`);
+    return state.version[addon.current_version.id];
+  }
+
+  return fetchVersionDetails({
```

Comment:
Yeah, I think https://github.com/mozilla/addons-server/issues/3964 will make all this code way simpler